torchvision/prototype/datasets/_builtin/coco.py

```diff
@@ -133,25 +177,121 @@ def _make_datapipe(
         anns_meta_dp = Mapper(anns_meta_dp, getitem(1))
         anns_meta_dp = UnBatcher(anns_meta_dp)
 
-        anns_dp = Grouper(anns_meta_dp, group_key_fn=getitem("image_id"), buffer_size=INFINITE_BUFFER_SIZE)
-        # drop images without annotations
-        anns_dp = Filter(anns_dp, bool)
-        anns_dp = Shuffler(anns_dp, buffer_size=INFINITE_BUFFER_SIZE)
-        anns_dp = IterKeyZipper(
-            anns_dp,
+        partial_anns_dp = Grouper(anns_meta_dp, group_key_fn=getitem("image_id"), buffer_size=INFINITE_BUFFER_SIZE)
+
+        return IterKeyZipper(
+            partial_anns_dp,
             images_meta_dp,
             key_fn=getitem(0, "image_id"),
             ref_key_fn=getitem("id"),
             buffer_size=INFINITE_BUFFER_SIZE,
         )
 
-        images_dp = ZipArchiveReader(images_dp)
+    def _precollate_anns(
+        self, data: List[Tuple[List[Dict[str, Any]], Dict[str, Any]]], *, types: List[str]
+    ) -> Tuple[Dict[str, Any], Dict[str, Dict[str, Any]]]:
+        ann_data, (image_meta, *_) = zip(*data)
+        anns = dict(zip(types, ann_data))
+        for empty_ann_types in set(self._ANN_TYPES) - set(types):
+            anns[empty_ann_types] = []
+        return image_meta, anns
 
-        dp = IterKeyZipper(
-            anns_dp,
-            images_dp,
-            key_fn=getitem(1, "file_name"),
-            ref_key_fn=path_accessor("name"),
+    def _make_anns_dp(
+        self, meta_dp: IterDataPipe[Tuple[str, io.IOBase]], *, config: DatasetConfig
+    ) -> IterDataPipe[Tuple[str, Dict[str, Dict[str, Any]]]]:
+        meta_dp = ZipArchiveReader(meta_dp)
+
+        ann_types = [type for type in self._ANN_TYPES if config[type]]
+        partial_anns_dps = Demultiplexer(
+            meta_dp,
+            len(ann_types),
+            functools.partial(
+                self._classifiy_meta_files,
+                split=config.split,
+                year=config.year,
+                ann_type_idcs=dict(zip(ann_types, range(len(ann_types)))),
+            ),
+            drop_none=True,
             buffer_size=INFINITE_BUFFER_SIZE,
         )
+        partial_anns_dps = [self._make_partial_anns_dp(dp) for dp in partial_anns_dps]
+
+        anns_dp = Concater(*partial_anns_dps)
+        anns_dp = Grouper(anns_dp, group_key_fn=getitem(1, "id"), buffer_size=INFINITE_BUFFER_SIZE)
```

Comment:
We should use `IterKeyZipper` here as we know the size of each group. If each datapipe in `partial_anns_dps` has a similar order, the `Grouper` would have to fill out most of buffer before yields any data.

Reply:
> We should use `IterKeyZipper` here as we know the size of each group. 

We don't. The grouper groups on the image id so basically this groups all annotations of one image together. So we neither know which annotations (instances, captions, ...) are present at all nor how many of them.

> If each datapipe in `partial_anns_dps` has a similar order, the `Grouper` would have to fill out most of buffer before yields any data.

Since we ride with an infinite buffer size, won't it fill up the entire buffer regardless of the order? In any case, I'm not sure why this is such bottleneck. I mean we are talking about roughly 500 MB of annotations and the key function is only item access. Surely that should be faster than 20 minutes, right? Reading the whole file into memory with `json.load` takes less than a second.

Reply:
It did some more benchmarking and it turns out that `Grouper` is indeed the bottleneck, but not at the place where you suspected it: the bottleneck happens when grouping all annotations of one type together.

I've benchmarked with [this script](https://gist.github.com/pmeier/6bf44e22f29f0bc840f8b9ba6a7f3830), which does the following:

1. Load one annotations json, i.e. one annotation type like instances or captions, from a zip archive.
2. Group all annotations of the same image id together.
3. Match the grouped annotations with the corresponding meta information of the image.

These steps are performed once without datapipes and once with. This is the result:

```
without datapipes 9.711 seconds
with datapipes 520.147 seconds
```

The `Grouper` seems to be hella inefficient.

Reply:
> We don't. The grouper groups on the image id so basically this groups all annotations of one image together. So we neither know which annotations (instances, captions, ...) are present at all nor how many of them.

IIUC, we have fixed number of `partial_anns_dps`. In each `partial_anns_dp`, the annotations have been aggregated into a list by the `Grouper` per image. Then, we know there are a certain amount of types (same as the number of `partial_anns_dp`). 

I will run the script and try to figure out if there is any work around.

Reply:
Thank to your script, I found the perf regression is caused by `Grouper` here:
https://github.com/pytorch/pytorch/blob/4eb772fde619c52f5877b917419beb8220160588/torch/utils/data/datapipes/iter/grouping.py#L318-L321

We are yielding group of contents with the order from bigger group to smaller group. Then, each iteration would take O(N) to find the biggest group. In total, it would be O(N^2) for whole epoch.

I will open an issue for it.

Reply:
I'm not sure what the initial motivation for returning ordered groups was in the first place, but that seems like a hidden feature. The docstring doesn't mention it and I wouldn't expect it either. IMO, this should be completely removed from the `Grouper` and if someone needs this sorting by group size, it can be easily added by a successive datapipe.

Reply:
Running with this patch

```diff
diff --git a/torch/utils/data/datapipes/iter/grouping.py b/torch/utils/data/datapipes/iter/grouping.py
index 37dcc6238b..ffc038232e 100644
--- a/torch/utils/data/datapipes/iter/grouping.py
+++ b/torch/utils/data/datapipes/iter/grouping.py
@@ -315,7 +315,5 @@ class GrouperIterDataPipe(IterDataPipe[DataChunk]):
                 if result_to_yield is not None:
                     yield self.wrapper_class(result_to_yield)
 
-        while buffer_size:
-            (result_to_yield, buffer_size) = self._remove_biggest_key(buffer_elements, buffer_size)
-            if result_to_yield is not None:
-                yield self.wrapper_class(result_to_yield)
+        for key in tuple(buffer_elements.keys()):
+            yield self.wrapper_class(buffer_elements.pop(key))
```

now gives

```
without datapipes 9.781 seconds
with datapipes 10.549 seconds
```

which sounds reasonable.

Reply:
Yeah. This is used for the case that buffer is smaller than the dataset, we would yield the biggest group. Even though it definitely can be optimized to O(log(n)) IMO.
Anyway, I will send a patch today.